File: Number_Theory/Climbing_Stairs.cpp
typedef struct {
    long long v11, v12, v21, v22;
} matrix;

matrix identitiy() {
    return matrix {1, 0, 1, 1};
}

matrix mul(matrix a, matrix b) {
    return matrix {a.v11*b.v11+a.v12*b.v21,
                   a.v11*b.v21+a.v12*b.v22,
                   a.v21*b.v11+a.v22*b.v21,
                   a.v21*b.v21+a.v22*b.v22};
}

matrix pow(matrix a, int b) {
    if(b==1)    return a;
    matrix ans = pow(a, b/2);
    ans = mul(ans, ans);
    if(b%2!=0)  ans = mul(ans, a);
    return ans;
}

class Solution {
public:
    int climbStairs(int n) {
        matrix fib = matrix{1, 1, 1, 0};
        if(n==1) return 1;
        if(n==2) return 2;
        matrix fibn = pow(fib, n-2);
        return 2*fibn.v11 + fibn.v12;
    }
};

File: Number_Theory/Climbing_Stairs.py

def mul(a, b):
    c = []
    for i in range(len(a)):
        row = []
        for j in range(len(b)):
            row.append(0)
            for k in range(len(a)):
                row[-1] += a[i][k]*b[k][j]
        c.append(row)
    return c
def pow(a, b):
    if b==0:
        return [[1, 0], [0, 1]]
    ans = pow(a, b//2)
    ans = mul(ans, ans)
    if b%2!=0:
        ans = mul(ans, a)
    return ans
class Solution:
    def climbStairs(self, n: int) -> int:
        m = [[1,1],[1,0]]
        if n==1:
            return 1
        if n==2:
            return 2
        m = pow(m, n-2)
        return 2*m[0][0]+m[0][1] 
